README.md
# plug-message-split

[![Greenkeeper badge](https://badges.greenkeeper.io/miniplug/plug-message-split.svg)](https://greenkeeper.io/)

Reliable message splitting for plug.dj chat messages.

## Why?

plug.dj's chat server truncates chat messages at 250 bytes (not characters). It also
first html-escapes some characters before truncating. This leads to funny bugs when
using certain characters:

![screenshot](error.png)

This library splits or truncates messages taking all known plug.dj oddities into
account.

## Installation

```sh
npm install --save plug-message-split
```

## API

```js
var msg = require('plug-message-split')
msg.split(string)
msg.splitRaw(string)
msg.truncate(string)
msg.escape(string)
msg.unescape(string)
```

### split(string): Array&lt;string>

Split a message for plug.dj on full words only.

### splitRaw(string): Array&lt;string>

Split a message on full characters. This might break words in two, but unicode
characters will be kept intact.

### truncate(string): string

Truncate a message to only full characters. Any content beyond the 250 bytes
limit is discarded. It might break words, but not characters.

### escape(string): string

HTML-escape a string like plug.dj escapes chat messages.

### unescape(string): string

Reverse plug.dj's chat message escaping. Useful for normalizing incoming chat
messages.

## License

[MIT]

[MIT]: ./LICENSE

CHANGELOG.md
# plug-message-split change log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## 2.0.0
* Support loading as a native ES module.


test.js
var assert = require('tapsert')
var m = require('./index')

assert.strictEqual(m.splitChars('Short and sweet').length, 1, 'Should not split short messages')
assert.strictEqual(m.splitWords('Short and sweet').length, 1, 'Should not split short messages')

// Hangul characters are grouped by syllable, but each syllable contains
// multiple letters and multiple bytes.
// plug.dj splits on the byte level, so it would cut off the last 안영하세요
// below in the middle of the 세 character.
var hangulStr = '안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 ' +
                '안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 ' +
                '안영하세요 안영하세요'

var result = m.splitChars(hangulStr)
assert.strictEqual(result[0],
  '안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 ' +
  '안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 ' +
  '안영하세요 안영하',
  'Should cut off before the 세 character'
)
assert.strictEqual(result[1],
  '세요',
  'Should put the remaining full characters in a new chunk'
)

result = m.splitWords(hangulStr)
assert.strictEqual(result[0],
  '안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 ' +
  '안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 안영하세요 ' +
  '안영하세요',
  'Should cut off before the last word and trim ending whitespace'
)
assert.strictEqual(result[1],
  '안영하세요',
  'Should put the remaining words in a new chunk'
)

// 63 >s are escaped to &gt;, and then cut off in the middle of the last
// escaped character. After unescaping, the message would look like:
// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>&g
var gtStr = '>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>'
assert.strictEqual(
  m.truncate(gtStr).length,
  62,
  'Should truncate before the last >'
)


index.mjs
export var MAX_LENGTH = 250

/**
 * HTML-escape a string like plug.dj escapes chat messages.
 */
export function escape (message) {
  return message
    .replace(/"/g, '&#34;')
    .replace(/'/g, '&#39;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/&/g, '&amp;')
}

/**
 * Reverse plug.dj's chat message escaping.
 */
export function unescape (message) {
  return message
    .replace(/&#34;/g, '"')
    .replace(/&#39;/g, '\'')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
}

/**
 * Get the size in bytes of a character in a plug.dj chat message.
 */
function byteLength (c) {
  // Escaped characters
  if (c === 34 /* " */ || c === 39 /* ' */ || c === 38 /* & */) {
    return 5
  }
  if (c === 60 /* < */ || c === 62 /* > */) {
    return 4
  }

  // Multi-byte unicode characters
  if (c > 0x7f && c <= 0x7ff) {
    return 2
  }
  if (c > 0x7ff && c <= 0xffff) {
    return 3
  }
  return 1
}

/**
 * Truncate a string to fit inside a plug.dj chat message.
 */
export function truncate (string) {
  var byteSize = 0
  for (var i = 0; i < string.length; i += 1) {
    var c = string.charCodeAt(i)
    var charSize = byteLength(c)

    if (byteSize + charSize > MAX_LENGTH) {
      return string.slice(0, i)
    }

    byteSize += charSize
  }

  return string
}

/**
 * Split a string into chunks that fit inside a plug.dj chat message. Splits on
 * characters, so it will break up words.
 */
export function splitChars (string) {
  var parts = []
  // Character index delimiting the current chunk
  var start = 0
  // Size in bytes of the current chunk
  var byteSize = 0
  for (var i = 0; i < string.length; i += 1) {
    var c = string.charCodeAt(i)
    var charSize = byteLength(c)

    if (byteSize + charSize > MAX_LENGTH) {
      parts.push(string.slice(start, i))
      start = i
      byteSize = 0
    }

    byteSize += charSize
  }

  if (byteSize !== 0) {
    parts.push(string.slice(start))
  }

  return parts
}

/**
 * Split a string into chunks. Attempts to only split on whitespace.
 */
export function splitWords (string) {
  var words = string.split(/(\S+\s+)/g)
  var parts = []
  var part = ''
  var size = 0
  for (var i = 0; i < words.length; i += 1) {
    var word = words[i]

    var wordSize = 0
    for (var c = 0; c < word.length; c += 1) {
      wordSize += byteLength(word.charCodeAt(c))
    }

    if (size + wordSize > MAX_LENGTH) {
      parts.push(part.replace(/\s+$/, ''))
      part = ''
      size = 0
    }

    part += word
    size += wordSize
  }

  if (size !== 0) {
    parts.push(part)
  }

  return parts
}

// Deprecated aliases
export { splitChars as splitRaw, splitWords as split }


index.js
var MAX_LENGTH = 250

/**
 * HTML-escape a string like plug.dj escapes chat messages.
 */
function escape (message) {
  return message
    .replace(/"/g, '&#34;')
    .replace(/'/g, '&#39;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/&/g, '&amp;')
}

/**
 * Reverse plug.dj's chat message escaping.
 */
function unescape (message) {
  return message
    .replace(/&#34;/g, '"')
    .replace(/&#39;/g, '\'')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
}

/**
 * Get the size in bytes of a character in a plug.dj chat message.
 */
function byteLength (c) {
  // Escaped characters
  if (c === 34 /* " */ || c === 39 /* ' */ || c === 38 /* & */) {
    return 5
  }
  if (c === 60 /* < */ || c === 62 /* > */) {
    return 4
  }

  // Multi-byte unicode characters
  if (c > 0x7f && c <= 0x7ff) {
    return 2
  }
  if (c > 0x7ff && c <= 0xffff) {
    return 3
  }
  return 1
}

/**
 * Truncate a string to fit inside a plug.dj chat message.
 */
function truncate (string) {
  var byteSize = 0
  for (var i = 0; i < string.length; i += 1) {
    var c = string.charCodeAt(i)
    var charSize = byteLength(c)

    if (byteSize + charSize > MAX_LENGTH) {
      return string.slice(0, i)
    }

    byteSize += charSize
  }

  return string
}

/**
 * Split a string into chunks that fit inside a plug.dj chat message. Splits on
 * characters, so it will break up words.
 */
function splitRaw (string) {
  var parts = []
  // Character index delimiting the current chunk
  var start = 0
  // Size in bytes of the current chunk
  var byteSize = 0
  for (var i = 0; i < string.length; i += 1) {
    var c = string.charCodeAt(i)
    var charSize = byteLength(c)

    if (byteSize + charSize > MAX_LENGTH) {
      parts.push(string.slice(start, i))
      start = i
      byteSize = 0
    }

    byteSize += charSize
  }

  if (byteSize !== 0) {
    parts.push(string.slice(start))
  }

  return parts
}

/**
 * Split a string into chunks. Attempts to only split on whitespace.
 */
function split (string) {
  var words = string.split(/(\S+\s+)/g)
  var parts = []
  var part = ''
  var size = 0
  for (var i = 0; i < words.length; i += 1) {
    var word = words[i]

    var wordSize = 0
    for (var c = 0; c < word.length; c += 1) {
      wordSize += byteLength(word.charCodeAt(c))
    }

    if (size + wordSize > MAX_LENGTH) {
      parts.push(part.replace(/\s+$/, ''))
      part = ''
      size = 0
    }

    part += word
    size += wordSize
  }

  if (size !== 0) {
    parts.push(part)
  }

  return parts
}

exports.MAX_LENGTH = MAX_LENGTH
exports.escape = escape
exports.unescape = unescape
exports.truncate = truncate
exports.splitRaw = splitRaw
exports.split = split